### PR TITLE
Collect telemetry for #55

### PR DESCRIPTION
**Pull request by chrmarti**

### CODE REVIEW

**Comment by bpasero** [comment]
@chrmarti would it not make more sense to collect the telemetry at the place where we actually search for files?

Today the `openAnythingHandler` combines the results of file search and symbol search. For #55 we only care about the file search, so I suggest to move the telemetry into the `openFileHandler` which is the one responsible of returning the file search results.

Note that the same service for file search is being used both for quick open and full text search. If we improve the raw speed of file search it would also show up for full text searches. If you care about the performance of file search in general, maybe we should even put telemetry one level down where we call to node to find all files.

**Comment by chrmarti** [comment]
@bpasero I've placed this first telemetry point such that it reflects the user's perceived performance. Measurements on my machine suggest that up to half of the time is be spent sorting the results when there are 250k results returned from the file search. I'll investigate further and will add more telemetry along the way.

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts [hunk 1]
```diff
@@ -76,6 +86,8 @@ export class OpenAnythingHandler extends QuickOpenHandler {
 	}
 
 	public getResults(searchValue: string): TPromise<QuickOpenModel> {
+		const timerEvent = this.telemetryService.timedPublicLog('openAnything');
+		const startTime = timerEvent.startTime ? timerEvent.startTime.getTime() : Date.now();
```
Comment (bpasero) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: I find the check weird here, why would startTime be undefined if you got timerEvent from the timedPublicLog method?

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts [hunk 2]
```diff
@@ -244,7 +274,7 @@ export class OpenAnythingHandler extends QuickOpenHandler {
 		return null;
 	}
 
-	public getResultsFromCache(searchValue: string, range: IRange = null): QuickOpenEntry[] {
+	public getResultsFromCache(searchValue: string, range: IRange = null) {
```
Comment (bpasero) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: One of our coding principles is to always type the return value of a function, even if TS can infer it. So please add an interface for this method too.

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts [hunk 3]
```diff
@@ -308,7 +340,12 @@ export class OpenAnythingHandler extends QuickOpenHandler {
 			entry.setHighlights(labelHighlights, descriptionHighlights);
 		});
 
-		return viewResults;
+		return {
```
Comment (bpasero) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: Can we make this nicer? E.g. an interface which has one entry for results and another for all performance/telemetry relevant data? Maybe just a Tupel<2>?

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts [hunk 4]
```diff
@@ -103,7 +115,15 @@ export class OpenAnythingHandler extends QuickOpenHandler {
 		// Check Cache first
 		let cachedResults = this.getResultsFromCache(searchValue, searchWithRange ? searchWithRange.range : null);
 		if (cachedResults) {
-			return TPromise.as(new QuickOpenModel(cachedResults));
+			timerEvent.data = <ITimerEventData>{
```
Comment (bpasero) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: I find quite some similarities to the ITimerEventData further below in the file, can we extract a method and use it in both places to construct the ITimerEventData?

**Comment by bpasero** [comment]
Added some comments to the code.

It is true that sorting is performance intense, I guess most of the time is spend computing a score for the given result based on the search value (see https://github.com/Microsoft/vscode/blob/chrmarti/filesearch/src/vs/base/common/scorer.ts#L1). 

You should not only look at the sorting, I have seen workspaces that almost never returned from all the node.js file walking we do. A good one is the Chromium sources (https://www.chromium.org/developers/how-tos/get-the-code) on a Windows machine as a test case 👍

**Comment by chrmarti** [comment]
I have done a few measurements of the file search alone under OS X on the Chromium sources and the current implementation is close to the fastest plain-JavaScript implementation I found. I'll repeat the tests with Windows. The `find` command line tool is way faster on OS X, haven't found an pm package with a native implementation to compare with yet.

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts (hunk 1 above)
Comment (chrmarti) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: startTime is null when telemetry is turned off during local development.

File: src/vs/workbench/parts/search/browser/openAnythingHandler.ts (hunk 3 above)
Comment (chrmarti) [inline src/vs/workbench/parts/search/browser/openAnythingHandler.ts]: Tuple types! Learned something. 👍

**Comment by bpasero** [comment]
I did play around with using find on Mac, Linux and the related command on Windows and in the end could not get much of a performance gain (it somewhat makes sense because in the end none of these commands leverage a search index). To me there are only two solutions to the slowness:
* either a real search index (e.g. Lucene)
* or maybe keep all file names in memory and then perform the search on this memory (that is what Atom does) and update it from file events

I am out on vacation starting tomorrow, I suggest you play around on this topic to gather some numbers and we connect when I am back.

**Comment by bpasero** [comment]
@chrmarti I was able to find my exploration where I used different variants of file search (including one where I spawn the native `find` command and `dir` on windows). see `recurseWithNativeCommand`  on https://gist.github.com/bpasero/fed45508c02ef35293d068c23b721d4b

**Comment by Tyriar** [comment]
Something I brought up that could lead somewhere is a faster recursive directory search using `tree` to build the index. This command (on all platforms) gets you all sub files and directories very fast, provided it can be parsed relatively quickly it may provide a speed up over a regular nodejs recursive stat to get the same information.

**Comment by bpasero** [comment]
@Tyriar don't forget that we do support excludes. If those native commands are unable to ignore specific large folders (node_modules etc.), you might end up with a longer search compared to what we do.

Btw we should also look into what Core CLR provides. I remember that C# has a good file walking API and it might be cross platform available now. We already use a C# file watcher on Windows, we could use a C# file walker. I actually found my old (kinda stupid) implementation for it: https://gist.github.com/bpasero/6fa8f2e0009ab8c245b9a6b1d96903ee

**Pull request merged**